pkg/container-collection/container-collection.go

```diff
@@ -119,16 +119,23 @@ func (cc *ContainerCollection) GetContainer(id string) *Container {
 	return container
 }
 
-// RemoveContainer removes a container from the collection.
+// RemoveContainer removes a container from the collection, but only after
+// notifying all the subscribers.
 func (cc *ContainerCollection) RemoveContainer(id string) {
-	v, loaded := cc.containers.LoadAndDelete(id)
+	v, loaded := cc.containers.Load(id)
 	if !loaded {
 		return
 	}
 
 	if cc.pubsub != nil {
 		cc.pubsub.Publish(EventTypeRemoveContainer, v.(*Container))
 	}
+
+	// Remove the container from the collection after publishing the event as
+	// subscribers might need to use the different collection's lookups during
+	// the notification handler, and they expect the container to still be
+	// present.
+	cc.containers.Delete(id)
 }
```

Comment:
I think it's fine to assume that `RemoveContainer()` and `AddContainer()` won't be called with the same container ID concurrently, otherwise a race can happen:

```
Goroutine A                    | Goroutine B 
                               |
RemoveContainer("foo")         | AddContainer("foo")
                               | ...
                               |   cc.containers.LoadOrStore("foo", container)
                               |   ^^^ -> loads existing container 
...                            |
  cc.containers.Delete("foo")  | 
  ^^^ -> removes container     |
  
  
...
GetContainer("foo") -> fails as container is not present
```